tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.runtimes.tomcat.session;
+
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.BlobValue;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.catalina.Manager;
+import org.apache.catalina.session.StandardSession;
+
+/**
+ * A DatastoreSession have the same behavior as a standard session but provide utilities to interact
+ * with the Datastore, such as helper for attributes and metadata serialization.
+ */
+public class DatastoreSession extends StandardSession {
+
+  protected Set<String> accessedAttributes;
+  protected Set<String> initialAttributes;
+
+  @VisibleForTesting
+  enum SessionMetadata {
```

Comment:
I might suggest using a separate class, rather than an enum to store the session metadata key names. I would do something like:
```
public class SessionMetadataConstants {
  public static final String CREATION_TIME_KEY = "creationTime";   
  // ... etc
}
```
I think it also would make sense for this class to be defined as a public class in its own separate file, since it's used by multiple other classes.

Reply:
I replaced the enumeration with a class, however I still think that it make sense to keep the class package private as most of the session attributes listed in that class are only accessible from within the session class (most of them have a protected visibility).

Reply:
What's the problem with the `enum`?